table.cc

```diff
@@ -180,7 +180,11 @@ table::make_reader(schema_ptr s,
         readers.emplace_back(mt->make_flat_reader(s, permit, range, slice, pc, trace_state, fwd, fwd_mr));
     }
 
-    if (cache_enabled() && !slice.options.contains(query::partition_slice::option::bypass_cache)) {
+    if (cache_enabled() && !slice.options.contains(query::partition_slice::option::bypass_cache)
+            && !slice.options.contains(query::partition_slice::option::reversed)) {
+        // FIXME: reenable cache for reversed reads after:
+        // - support for reversed reads is implemented in the cache,
+        // - fast forwarding is implemented in reversed sstable readers.
         readers.emplace_back(_cache.make_reader(s, permit, range, slice, pc, std::move(trace_state), fwd, fwd_mr));
```

Comment:
This is introducing a regression for users that have well-cached small reversed reads.

Let's add a configuration item (LiveUpdate::yes) to select whether we want to go through the cache or not. It can default to the current behavior, and users who experience problems with larger partitions can enable it. Once cache supports reversed reads we can drop it.

Since the old reversing reader will provide cache with the table schema, it will send non-reversed reads to the sstable reader and things should work out.

Reply:
> Let's add a configuration item (LiveUpdate::yes) to select whether we want to go through the cache or not. It can default to the current behavior, and users who experience problems with larger partitions can enable it. Once cache supports reversed reads we can drop it.

Please confirm if I understand this correctly. We want a config option such that (switch enabled/disabled if necessary):
- if it's enabled, we use the new reversing method implemented inside sstables and disable the cache,
- if it's disabled, we perform a forward query in the sstable, enable the cache, and then reverse the result returned from the cache using `make_reversing_reader`,
- the option is disabled by default?

> Since the old reversing reader will provide cache with the table schema, it will send non-reversed reads to the sstable reader and things should work out.

I'm not sure I understand. Please elaborate.

Reply:
> > Let's add a configuration item (LiveUpdate::yes) to select whether we want to go through the cache or not. It can default to the current behavior, and users who experience problems with larger partitions can enable it. Once cache supports reversed reads we can drop it.
> 
> Please confirm if I understand this correctly. We want a config option such that (switch enabled/disabled if necessary):
> 
> * if it's enabled, we use the new reversing method implemented inside sstables and disable the cache,
> * if it's disabled, we perform a forward query in the sstable, enable the cache, and then reverse the result returned from the cache using `make_reversing_reader`,

(the current status quo)

> * the option is disabled by default?

Yes.

> 
> > Since the old reversing reader will provide cache with the table schema, it will send non-reversed reads to the sstable reader and things should work out.
> 
> I'm not sure I understand. Please elaborate.

Status-quo reads with new code:

table.cc:
1. detect query order != table order
2. decompose into (a) forward query and (b) make_reversed_reader(). Sub-query (a) has query order = table order

row_cache.cc:
1. Process forward query, if miss call into sstable reader

sstable reader:
It's a normal forward query, so the new reversing code doesn't trigger.